crates/host/src/guest.rs

```diff
@@ -36,15 +35,57 @@ use wasmer_runtime::Value;
 /// @see https://docs.rs/wasmer-runtime-core/0.17.0/src/wasmer_runtime_core/memory/ptr.rs.html#120
 ///
 /// this is still not completely safe in the face of shared memory and threads, etc.
-pub fn write_slice(ctx: &mut Ctx, guest_ptr: RemotePtr, slice: &[u8]) {
+///
+/// the guest needs to provide a pointer to a pre-allocated (e.g. by forgetting a Vec<u8>) region
+/// of the guest's memory that it is safe for the host to write to.
+///
+/// it is the host's responsibility to tell the guest the length of the allocation that is needed
+/// and the guest's responsibility to correctly reserve an allocation to be written into.
+///
+/// write_bytes() takes a slice of bytes and writes it to the position at the guest pointer
+///
+/// as the byte slice cannot be co-ordinated by the compiler (because the host and guest have
+/// different compilers and allocators) we prefix the allocation with a WasmSize length value.
+///
+/// for example, if we wanted to write the slice &[1, 2, 3] then we'd take the length of the slice,
+/// 3 as a WasmSize, which is u32, i.e. a 3_u32 and convert it to an array of u8 bytes as
+/// [ 3_u8, 0_u8, 0_u8, 0_u8 ] and concatenate it to our original [ 1_u8, 2_u8, 3_u8 ].
+/// this gives the full array of bytes to write as:
+///
+/// ```ignore
+/// [ 3_u8, 0_u8, 0_u8, 0_u8, 1_u8, 2_u8, 3_u8 ]
+/// ```
+///
+/// this allows us to read back the byte slice given only a GuestPtr because the read operation
+/// can do the inverse in a single step by reading the length inline
+///
+/// it also requires the host and the guest to both adopt this convention and read/write the
+/// additional 4 byte prefix in order to read/write the real payload correctly
+///
+/// @see read_bytes()
+pub fn write_bytes(ctx: &mut Ctx, guest_ptr: GuestPtr, slice: &[u8]) -> Result<(), WasmError> {
     let ptr: WasmPtr<u8, Array> = WasmPtr::new(guest_ptr as _);
-    for (byte, cell) in slice.iter().zip(unsafe {
-        ptr.deref_mut(ctx.memory(0), 0, slice.len() as _)
-            .expect("pointer in bounds")
-            .iter()
-    }) {
+
+    // build the length prefix slice
+    let len = slice.len() as Len;
+    let len_bytes: [u8; 4] = len.to_le_bytes();
```

Comment:
Everywhere else the `WasmSize` is well abstracted - you could do a

```rust
let len_bytes = vec![0; sizeof(WasmSize)];
```

here... but probably YAGNI as wasm64 is a long way off still.